opentelemetry/src/main/scala/zio/telemetry/opentelemetry/Tracing.scala

```diff
@@ -98,6 +104,77 @@ object Tracing {
       r   <- createChildOf(old, spanName, spanKind).use(finalizeSpanUsingEffect(effect, _, toErrorStatus))
     } yield r
 
+  /*
+   * Introduces a thread-local scope during the execution allowing
+   * for non-zio context propagation.
+   *
+   * Closes the scope when the effect finishes.
+   */
+  def scopedEffect[R, A](effect: => A): ZIO[Tracing, Throwable, A] =
+    for {
+      currentSpan <- getCurrentSpan
+      tracer      <- getTracer
+      eff <- Task.effect {
+              // TODO: best practice is to use java try-with-resources. Is there
+              // something from ZIO or stdlib we can use that handles errors in close?
+              val scope = tracer.withSpan(currentSpan)
+              try {
+                effect
+              } finally {
+                scope.close()
+              }
+            }
+    } yield eff
+
+  /*
+   * Introduces a thread-local scope during the execution allowing
+   * for non-zio context propagation.
+   *
+   * Closes the scope when the effect finishes.
+   */
+  def scopedEffectTotal[R, A](effect: => A): ZIO[Tracing, Nothing, A] =
+    for {
+      currentSpan <- getCurrentSpan
+      tracer      <- getTracer
+      eff <- Task.effectTotal {
+              // TODO: best practice is to use java try-with-resources. Is there
+              // something from ZIO or stdlib we can use that handles errors in close?
+              val scope = tracer.withSpan(currentSpan)
+              try {
+                effect
+              } finally {
+                scope.close()
+              }
+            }
+    } yield eff
+
+  /*
+   * Introduces a thread-local scope from the currently active zio span
+   * allowing for non-zio context propagation. This scope will only be
+   * active during Future creation, so another mechanism must be used to
+   * ensure that the scope is passed into the Future callbacks.
+   *
+   * The java auto instrumentation package provides such a mechanism out of
+   * the box, so one is not provided as a part of this method.
+   *
+   * CLoses the scope when the effect finishes
+   */
+  def scopedEffectFromFuture[R, A](make: ExecutionContext => scala.concurrent.Future[A]): ZIO[Tracing, Throwable, A] =
+    for {
+      currentSpan <- getCurrentSpan
+      tracer      <- getTracer
+      eff <- ZIO.fromFuture { implicit ec =>
+              // TODO: best practice is to use java try-with-resources. Is there
+              // something from ZIO or stdlib we can use that handles errors in close?
```

Comment:
How about `ZManaged` around the `scope`?

Reply:
I considered ZManaged, but I haven't been able to figure out how to use it without tying the lifetime to the zio runtime instead of the local block.  I think technically a "bracket" does what we need here but again, it's tightly tied to the ZIO and I haven't figured out the right way to unwind it.

I do have fairly good evidence that the current approach works so given this works I can take another look at optimizing it.

Reply:
According to the JavaDoc, close cannot throw an `Exception`, so I guess the current code should be fine.

Reply:
I've pushed an update without the comments and tightened up the try-finally. Given close doesn't throw, as you say, I think keeping it simple and matching the upstream recommendation is probably better than trying to handle with with zmanaged or bracket and having to rely on the zio runtime to ensure everything runs on the correct thread.